File: src/cli/commands/hooks.ts
import { chmodSync, existsSync, mkdirSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import type { Command } from "commander";
import { Registry } from "../../core/registry.js";
import { log } from "../../utils/logger.js";

const POST_COMMIT_HOOK = `#!/bin/sh
# graphmind: auto-rebuild graph after commit
graphmind build 2>/dev/null &
`;

const PRE_PUSH_HOOK = `#!/bin/sh
# graphmind: show diff-impact before push
echo ""
echo "  graphmind diff-impact:"
graphmind diff-impact 2>/dev/null
echo ""
`;

function installHook(hooksDir: string, name: string, content: string): boolean {
	const hookPath = join(hooksDir, name);

	if (existsSync(hookPath)) {
		const existing = readFileSync(hookPath, "utf-8");
		if (existing.includes("graphmind")) {
			log.dim(`  ${name}: already installed`);
			return false;
		}
		writeFileSync(hookPath, `${existing}\n${content}`);
	} else {
		writeFileSync(hookPath, content);
	}

	chmodSync(hookPath, 0o755);
	return true;
}

export function registerHooksCommand(program: Command): void {
	const hooks = program.command("hooks").description("Manage git hooks for graphmind");

	hooks
		.command("install [slug]")
		.description("Install git hooks (post-commit + pre-push)")
		.action((slug?: string) => {
			const registry = new Registry();
			const project = slug ? registry.get(slug) : registry.findByPath(process.cwd());

			if (!project) {
				log.error("Not in a registered project.");
				process.exitCode = 1;
				return;
			}

			const gitDir = join(project.path, ".git");
			if (!existsSync(gitDir)) {
				log.error(`No .git directory in ${project.path}`);
				process.exitCode = 1;
				return;
			}

			const hooksDir = join(gitDir, "hooks");
			mkdirSync(hooksDir, { recursive: true });

			let installed = 0;
			if (installHook(hooksDir, "post-commit", POST_COMMIT_HOOK)) installed++;
			if (installHook(hooksDir, "pre-push", PRE_PUSH_HOOK)) installed++;

			if (installed > 0) {
				log.success(`Installed ${installed} hook(s) for "${project.slug}"`);
			} else {
				log.dim("All hooks already installed.");
			}
		});

	hooks
		.command("uninstall [slug]")
		.description("Remove graphmind git hooks")
		.action((slug?: string) => {
			const registry = new Registry();
			const project = slug ? registry.get(slug) : registry.findByPath(process.cwd());

			if (!project) {
				log.error("Not in a registered project.");
				process.exitCode = 1;
				return;
			}

			const hooksDir = join(project.path, ".git", "hooks");
			let removed = 0;

			for (const name of ["post-commit", "pre-push"]) {
				const hookPath = join(hooksDir, name);
				if (!existsSync(hookPath)) continue;

				const content = readFileSync(hookPath, "utf-8");
				if (!content.includes("graphmind")) continue;

				const lines = content.split("\n");
				const filtered = lines.filter((l) => !l.includes("graphmind"));
				const remaining = filtered.filter((l) => l.trim() && l.trim() !== "#!/bin/sh").length;

				if (remaining === 0) {
					unlinkSync(hookPath);
				} else {
					writeFileSync(hookPath, filtered.join("\n"));
				}
				removed++;
			}

			if (removed > 0) {
				log.success(`Removed graphmind hooks from "${project.slug}"`);
			} else {
				log.dim("No graphmind hooks found.");
			}
		});
}

File: crates/graphmind-cli/src/commands/hooks.rs
use crate::config::Registry;
use crate::resolve::resolve_project_slug;
use colored::Colorize;
use std::fs;
use std::os::unix::fs::PermissionsExt;
use std::path::Path;

const POST_COMMIT_HOOK: &str = "#!/bin/sh\n# graphmind: auto-rebuild\ngraphmind build 2>/dev/null &\n";

const PRE_PUSH_HOOK: &str = "#!/bin/sh\n# graphmind: diff-impact\necho \"\"\necho \"  graphmind diff-impact:\"\ngraphmind diff-impact 2>/dev/null\necho \"\"\n";

fn install_hook(hooks_dir: &Path, name: &str, content: &str) -> bool {
    let hook_path = hooks_dir.join(name);

    if hook_path.exists() {
        let existing = fs::read_to_string(&hook_path).unwrap_or_default();
        if existing.contains("graphmind") {
            println!("  {}: already installed", name.dimmed());
            return false;
        }
        fs::write(&hook_path, format!("{existing}\n{content}")).ok();
    } else {
        fs::write(&hook_path, content).ok();
    }

    fs::set_permissions(&hook_path, fs::Permissions::from_mode(0o755)).ok();
    true
}

pub fn install(slug: Option<&str>) {
    let slug = match resolve_project_slug(&[slug]) {
        Some(s) => s,
        None => {
            eprintln!("{} Not in a registered project.", "Error:".red().bold());
            std::process::exit(1);
        }
    };

    let project = match Registry::get(&slug) {
        Some(p) => p,
        None => {
            eprintln!("{} Project \"{}\" not found.", "Error:".red().bold(), slug);
            std::process::exit(1);
        }
    };

    let git_dir = Path::new(&project.path).join(".git");
    if !git_dir.exists() {
        eprintln!("{} No .git directory in {}", "Error:".red().bold(), project.path);
        std::process::exit(1);
    }

    let hooks_dir = git_dir.join("hooks");
    fs::create_dir_all(&hooks_dir).ok();

    let mut installed = 0;
    if install_hook(&hooks_dir, "post-commit", POST_COMMIT_HOOK) {
        installed += 1;
    }
    if install_hook(&hooks_dir, "pre-push", PRE_PUSH_HOOK) {
        installed += 1;
    }

    if installed > 0 {
        println!(
            "{} Installed {} hook(s) for \"{}\"",
            "OK".green().bold(),
            installed,
            slug
        );
    } else {
        println!("{}", "All hooks already installed.".dimmed());
    }
}

pub fn uninstall(slug: Option<&str>) {
    let slug = match resolve_project_slug(&[slug]) {
        Some(s) => s,
        None => {
            eprintln!("{} Not in a registered project.", "Error:".red().bold());
            std::process::exit(1);
        }
    };

    let project = match Registry::get(&slug) {
        Some(p) => p,
        None => {
            eprintln!("{} Project \"{}\" not found.", "Error:".red().bold(), slug);
            std::process::exit(1);
        }
    };

    let hooks_dir = Path::new(&project.path).join(".git").join("hooks");
    let mut removed = 0;

    for name in &["post-commit", "pre-push"] {
        let hook_path = hooks_dir.join(name);
        if !hook_path.exists() {
            continue;
        }

        let content = match fs::read_to_string(&hook_path) {
            Ok(c) => c,
            Err(_) => continue,
        };

        if !content.contains("graphmind") {
            continue;
        }

        let filtered: Vec<&str> = content
            .lines()
            .filter(|l| !l.contains("graphmind"))
            .collect();

        let remaining = filtered
            .iter()
            .filter(|l| !l.trim().is_empty() && l.trim() != "#!/bin/sh")
            .count();

        if remaining == 0 {
            fs::remove_file(&hook_path).ok();
        } else {
            fs::write(&hook_path, filtered.join("\n")).ok();
        }
        removed += 1;
    }

    if removed > 0 {
        println!(
            "{} Removed graphmind hooks from \"{}\"",
            "OK".green().bold(),
            slug
        );
    } else {
        println!("{}", "No graphmind hooks found.".dimmed());
    }
}
